test: replace `ElasticsearchDocumentStore` with `InMemoryDocumentStore` when it makes sense

anakin87 opened the pull request:
### Related Issues
- fixes #4282

### Proposed Changes:
I noticed that several tests use `ElasticsearchDocumentStore` only to test BM25 retrieval.
I modified them, introducing `InMemoryDocumentStore` (since it supports BM25), to make them lighter.

### How did you test it?
CI

### Notes for the reviewer
Only a first draft

### Checklist
- [X] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [X] I have updated the related issue with new insights and changes
- [ ] I added tests that demonstrate the correct behavior of the change
- [X] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [ ] I documented my code
- [X] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

ZanSara commented:
Hey @anakin87 Thank you so much for this refactor already!! Please tag me or @masci to fast track the review as soon as you think it's ready :rocket:

anakin87 commented:
- replaced `ElasticsearchDocumentStore` with `InMemoryDocumentStore` in 12 tests
- set `bm25_parameters` in `InMemoryDS` to make it behave similar to `ElasticDS`
- slightly adjusted some scores in test
- added support for scaling scores in `InMemoryDS` (minimal and easy change)

- I still use `ElasticDS` in `test_eval_elastic_retriever`:
in fact, I have verified that when dealing with very few (2 in this case) and short documents,
`InMemoryDS` BM25 scores are meaningless (very close to zero)

anakin87 commented:
@ZanSara ready for review 🙂

ZanSara commented:
Hey @mayankjobanputra, what makes this PR a breaking change? This PR is mostly about tests.

ZanSara commented on test/conftest.py:
```diff
@@ -863,6 +863,7 @@ def get_document_store(
             index=index,
             similarity=similarity,
             use_bm25=True,
+            bm25_parameters={"k1": 1.2, "b": 0.75},  # parameters similar to those of Elasticsearch
```
Not a problem but I'm not sure we even need these. Is there any test where you found a large difference?

anakin87 replied:
Yes, I found some tests that compare the obtained scores with the expected ones and this helps a little.

ZanSara reviewed (approved):
Very nice, no remarks! Thank you for the help :rocket:

ZanSara commented:
@agnieszka-m: for the docs, this PR adds support for `scale_score` in `InMemoryDocumentStore`. All other changes are related to tests.

mayankjobanputra commented on haystack/document_stores/memory.py:
```diff
@@ -953,14 +954,15 @@ def query(
         headers: Optional[Dict[str, str]] = None,
         all_terms_must_match: bool = False,
-        scale_score: bool = False,
+        scale_score: bool = True,
```
@ZanSara This is the breaking change. Don't you think?

ZanSara replied:
Not really, the default should be True and it was False only because it was not supported

anakin87 replied:
Ah.
I aligned the parameters of this document store to the default values ​​for other similar document stores...

For example:
https://github.com/deepset-ai/haystack/blob/e8f9b1b65dc8a7f36e6e9bdcd75b177ff4019edd/haystack/document_stores/search_engine.py#L749
